File: src/jastudio_src/jastudio/language_services/janome_ex/word_extraction/analysis_constants.py
from __future__ import annotations

from jastudio.sysutils import ex_str

real_quote_characters: set[str] = {"「", "」", '"'}
pseudo_quote_characters: set[str] = {"と", "って"}
all_quote_characters: set[str] = real_quote_characters | pseudo_quote_characters

space_characters: set[str] = {" ", "\t", ex_str.invisible_space}

question_marks: set[str] = {"？", "?"}
periods: set[str] = {".", "。", "｡"}
commas: set[str] = {",", "、"}
tilde: set[str] = {"～", "~"}


exclamations: set[str] = {"!"} #todo: isn't there a full size exclamation mark?

all_punctuation_characters: set[str] = all_quote_characters | question_marks | periods | commas | exclamations | tilde | {":", ";", "/", "|"}


sentence_start_characters: set[str] = real_quote_characters | space_characters | question_marks | periods
sentence_end_characters: set[str] = all_quote_characters | space_characters | question_marks | periods
noise_characters = all_punctuation_characters | space_characters | pseudo_quote_characters | sentence_end_characters


passive_verb_endings: set[str] = {"あれる", "られる", "される"}
causative_verb_endings: set[str] = {"あせる", "させる", "あす", "さす"}

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/WordExtraction/AnalysisConstants.cs
using System.Collections.Generic;

namespace JAStudio.Core.LanguageServices.JanomeEx.WordExtraction;

public static class AnalysisConstants
{
   static readonly HashSet<string> RealQuoteCharacters = ["「", "」", "\""];
    public static readonly HashSet<string> PseudoQuoteCharacters = ["と", "って"];
    public static readonly HashSet<string> AllQuoteCharacters;

    public static readonly HashSet<string> SpaceCharacters = [" ", "\t", StringExtensions.InvisibleSpace];

    public static readonly HashSet<string> QuestionMarks = ["？", "?"];
    public static readonly HashSet<string> Periods = [".", "。", "｡"];
    public static readonly HashSet<string> Commas = [",", "、"];
    public static readonly HashSet<string> Tilde = ["～", "~"];
    public static readonly HashSet<string> Exclamations = ["!"]; // TODO: Full-width exclamation mark?

    public static readonly HashSet<string> AllPunctuationCharacters;
    public static readonly HashSet<string> SentenceStartCharacters;
    public static readonly HashSet<string> SentenceEndCharacters;
    public static readonly HashSet<string> NoiseCharacters;

    public static readonly HashSet<string> PassiveVerbEndings = ["あれる", "られる", "される"];
    public static readonly HashSet<string> CausativeVerbEndings = ["あせる", "させる", "あす", "さす"];

    static AnalysisConstants()
    {
        AllQuoteCharacters = new HashSet<string>(RealQuoteCharacters);
        AllQuoteCharacters.UnionWith(PseudoQuoteCharacters);

        AllPunctuationCharacters = new HashSet<string>(AllQuoteCharacters);
        AllPunctuationCharacters.UnionWith(QuestionMarks);
        AllPunctuationCharacters.UnionWith(Periods);
        AllPunctuationCharacters.UnionWith(Commas);
        AllPunctuationCharacters.UnionWith(Exclamations);
        AllPunctuationCharacters.UnionWith(Tilde);
        AllPunctuationCharacters.UnionWith([":", ";", "/", "|"]);

        SentenceStartCharacters = new HashSet<string>(RealQuoteCharacters);
        SentenceStartCharacters.UnionWith(SpaceCharacters);
        SentenceStartCharacters.UnionWith(QuestionMarks);
        SentenceStartCharacters.UnionWith(Periods);

        SentenceEndCharacters = new HashSet<string>(AllQuoteCharacters);
        SentenceEndCharacters.UnionWith(SpaceCharacters);
        SentenceEndCharacters.UnionWith(QuestionMarks);
        SentenceEndCharacters.UnionWith(Periods);

        NoiseCharacters = new HashSet<string>(AllPunctuationCharacters);
        NoiseCharacters.UnionWith(SpaceCharacters);
        NoiseCharacters.UnionWith(PseudoQuoteCharacters);
        NoiseCharacters.UnionWith(SentenceEndCharacters);
    }
}
